R/data.table.R

```diff
@@ -1739,22 +1739,47 @@ replace_dot_alias = function(e) {
         GForce = FALSE
       } else {
         # Apply GForce
+        noCall_noVars = function(expr) (!is.call(expr) || length(all.vars(expr, max.names=1L, unique=FALSE))==0L) && !dotN(expr)
+        .gshift_ok = function(q) {
+          q = match.call(shift, q)
+          noCall_noVars(q[["n"]]) &&
+            noCall_noVars(q[["fill"]]) &&
+            noCall_noVars(q[["type"]]) &&
+            is.null(q[["give.names"]])
+        }
+        .ghead_ok = function(q) {
+          length(q) == 3L &&
+            length(q[[3L]]) == 1L &&
+            noCall_noVars(q[[3L]])
+        }
+        ".g[_ok" = function(q, x) {
+          length(q) == 3L &&
+            length(q[[3L]]) == 1L &&
+            noCall_noVars(q[[3L]]) &&
+            (q[[1L]] == "[" || (q[[1L]] == "[[" && eval(call('is.atomic', q[[2L]]), envir=x))) &&
+            eval(q[[3L]], parent.frame(3L)) > 0L
+        }
+        .gweighted.mean_ok = function(q, x) { #3977
+          q = match.call(function(x, w, ..., na.rm=FALSE) {}, q)
+          noCall_noVars(q[["na.rm"]]) &&
+            (is.null(q[["w"]]) || eval(call('is.numeric', q[["w"]]), envir=x))
+        }
         .gforce_ok = function(q) {
           if (dotN(q)) return(TRUE) # For #334
           # run GForce for simple f(x) calls and f(x, na.rm = TRUE)-like calls where x is a column of .SD
           # is.symbol() is for #1369, #1974 and #2949
-          if (!(is.call(q) && is.symbol(q[[1L]]) && is.symbol(q[[2L]]) && (q1 <- q[[1L]]) %chin% gfuns)) return(FALSE)
+          if (!(is.call(q) && is.symbol(q[[1L]]) && is.symbol(q[[2L]]) && (q[[1L]]) %chin% gfuns)) return(FALSE)
           if (!(q2 <- q[[2L]]) %chin% names(SDenv$.SDall) && q2 != ".I") return(FALSE)  # 875
-          if ((length(q)==2L || (!is.null(names(q)) && startsWith(names(q)[3L], "na")))) return(TRUE)
+          if (length(q)==2L || (!is.null(names(q)) && startsWith(names(q)[3L], "na") && noCall_noVars(q[[3L]]))) return(TRUE)
           #                       ^^ base::startWith errors on NULL unfortunately
-          if (length(q)>=2L && q[[1L]] == "shift") {
-            q_named = match.call(shift, q)
-            if (!is.call(q_named[["fill"]]) && is.null(q_named[["give.names"]])) return(TRUE)
-          }
-          if (length(q)>=3L && q[[1L]] == "weighted.mean") return(TRUE)  #3977
-          # otherwise there must be three arguments
-          length(q)==3L && length(q3 <- q[[3L]])==1L && is.numeric(q3) &&
-            ( (q1 %chin% c("head", "tail")) || ((q1 == "[" || (q1 == "[[" && eval(call('is.atomic', q[[2L]]), envir=x))) && q3>0L) )
+          switch(as.character(q[[1L]]), 
+            "shift" = .gshift_ok(q),
+            "weighted.mean" = .gweighted.mean_ok(q, x),
+            "tail" = , "head" = .ghead_ok(q),
+            "[" = `.g[_ok`(q, x),
+            "[[" = `.g[_ok`(q, x),
```

Comment:
```suggestion
            "[[" = , "[" = `.g[_ok`(q, x),
```

OK, I searched for how to make multi-line suggestions:

https://github.blog/changelog/2020-02-26-multi-line-code-suggestions-beta/